apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/helper/DubboTraceHelper.java

```diff
@@ -0,0 +1,162 @@
+import co.elastic.apm.agent.impl.context.Destination;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+
+import java.net.InetSocketAddress;
+import java.net.URL;
+import java.security.CodeSource;
+import java.security.ProtectionDomain;
+
+@VisibleForAdvice
+public class DubboTraceHelper {
+
+    @VisibleForAdvice
+    public static ElasticApmTracer tracer;
+
+    private static String EXTERNAL_TYPE = "external";
+
+    private static String DUBBO_SUBTYPE = "dubbo";
+
+    public static final String PROVIDER_SERVICE_NAME_KEY = "elastic-apm-dubbo-provider";
+
+    public static void init(ElasticApmTracer tracer) {
+        DubboTraceHelper.tracer = tracer;
+    }
+
+    @VisibleForAdvice
+    public static String buildDubboRequestName(DubboApiInfo apiInfo) {
+        Class<?>[] paramClasses = apiInfo.getParamClasses();
+        String paramsSign = "";
+        if (paramClasses != null && paramClasses.length > 0) {
+            StringBuilder paramSignBuilder = new StringBuilder(paramClasses[0].getSimpleName());
+            for (int i = 1; i < paramClasses.length; i++) {
+                paramSignBuilder.append(",").append(paramClasses[i].getSimpleName());
+            }
+            paramsSign = paramSignBuilder.toString();
+        }
+
+        String requestName = apiInfo.getApiClass().getName() + "." + apiInfo.getMethodName() + "(" + paramsSign + ")";
+        String version = apiInfo.getVersion();
+        if (version != null && version.length() > 0) {
+            requestName += " version=" + version;
+        }
+
+        return requestName;
+    }
+
+    @VisibleForAdvice
+    public static Span createConsumerSpan(DubboApiInfo apiInfo, InetSocketAddress remoteAddress) {
+        TraceContextHolder<?> traceContext = DubboTraceHelper.tracer.getActive();
+        if (traceContext == null) {
+            return null;
+        }
+        Span span = traceContext.createExitSpan();
+        if (span == null) {
+            return null;
+        }
+
+        span.withType(EXTERNAL_TYPE)
+            .withSubtype(DUBBO_SUBTYPE)
+            .withName(buildDubboRequestName(apiInfo));
+        Destination destination = span.getContext().getDestination();
+        destination.withAddress(remoteAddress.getHostName()).withPort(remoteAddress.getPort());
+
+        Destination.Service service = destination.getService();
+        service.withType(EXTERNAL_TYPE).withResource(DUBBO_SUBTYPE).withName(DUBBO_SUBTYPE);
+
+        return span.activate();
+    }
+
+    @VisibleForAdvice
+    public static void fillTransaction(Transaction transaction, DubboApiInfo dubboApiInfo) {
+        transaction.withName(buildDubboRequestName(dubboApiInfo));
+        transaction.withType("dubbo");
+        transaction.activate();
+    }
+
+    public static boolean isBizException(Class<?> interfaceClass, Class<?> exp) {
```

Comment:
Could you explain what this exactly does? I fear that this is going to have too much runtime overhead.

Reply:
the best pratice by dubbo, we will throw some exception that we expected, we call it as the business exception, so we don't want to let these  exceptions to be shown as error in kibana.
 
one way is ignoring these exception by setting the CoreConfigurations, but it's inconvenient to settings the configuration for each application， because each application has their own business exceptions, so i decide to handle it dynamically.

for performance, one exception class will only be judged whether it's business exception once, and the agent will cache the result.

Reply:
You are already trying to capture business exceptions with custom attributes. I would still just capture a business exception just like a regular one. Then you can both see business exceptions and actual errors in the UI.

Could you make an example of when you would throw a business exception? Would one example be someone trying to order an item when there's no stock for that item? Like `NoToiletPaperException`?

I would argue that it's still interesting to monitor how many of these exceptions are thrown over time.

Reply:
ok, You are right.

Reply:
Are you planning to remove the biz exception concept?